src/modifierSelfImmunity.ts

```diff
@@ -0,0 +1,42 @@
+import { registerEvents, registerModifiers } from "./cards";
+import { getOrInitModifier } from "./cards/util";
+import * as Unit from './entity/Unit';
+import Underworld from './Underworld';
+
+// Grants a unit invulnerability to its own damage (I.E. explosives)
+export const selfInvulnerabilityId = 'Self Invulnerability';
+export default function registerSelfInvulnerability() {
+  registerModifiers(selfInvulnerabilityId, {
+    description: 'Grants a unit invulnerability to its own damage (I.E. explosives)',
+    add: (unit: Unit.IUnit, underworld: Underworld, prediction: boolean, quantity: number = 1) => {
+      getOrInitModifier(unit, selfInvulnerabilityId, { isCurse: false, quantity, keepOnDeath: true }, () => {
+        Unit.addEvent(unit, selfInvulnerabilityId);
+      });
+
+      if (!prediction) {
+        updateTooltip(unit);
+      }
+    }
+  });
+  registerEvents(selfInvulnerabilityId, {
+    onTakeDamage: (unit: Unit.IUnit, amount: number, underworld: Underworld, prediction: boolean, damageDealer?: Unit.IUnit) => {
+      const modifier = unit.modifiers[selfInvulnerabilityId];
+      if (modifier) {
+        // If the source of incoming damage is myself, negate it
+        if (damageDealer == unit && amount > 0) {
+          amount = 0;
```

Comment:
We try to "show" when a modifier takes effect (via VFX or floating text) so there isn't any mystery as to why something happened.